### PR TITLE
fix(semantic/jsdoc): Support multibyte chars

### PR DESCRIPTION
**Pull request by leaysgur**
Fix for https://github.com/oxc-project/oxc/pull/2642#issuecomment-1985258633

Any advice would be appreciated as I'm not too sure about this area. 🥹

### CODE REVIEW

File: crates/oxc_semantic/src/jsdoc/parser/parse.rs [hunk 1]
```diff
@@ -35,14 +35,14 @@ impl<'a> JSDocParser<'a> {
         let mut tags = vec![];
 
         // Let's start with the first `@`
-        while let Some(c) = self.source_text.chars().nth(self.current) {
+        while let Some((_, c)) = self.source_text[self.current..].char_indices().next() {
             match c {
                 '@' => {
-                    self.current += 1;
+                    self.current += c.len_utf8();
```
Comment (Boshen) [inline crates/oxc_semantic/src/jsdoc/parser/parse.rs:41]: I think you can write `self.current = i` given `char_indices` gives you the character offset.

File: crates/oxc_semantic/src/jsdoc/parser/parse.rs [hunk 2]
```diff
@@ -445,4 +459,61 @@ comment */"
...
+                    kind: JSDocTagKind::Parameter(Param {
+                        name: "getRowKey",
+                        r#type: Some(ParamType { value: "GetRowKey<T>" })
+                    }),
+                    comment: "获取当前rowKey的方法".to_string(),
+                },
+                JSDocTag {
+                    kind: JSDocTagKind::Unknown("returns"),
+                    comment: "flattened data".to_string(),
+                },
```
Comment (Boshen) [inline crates/oxc_semantic/src/jsdoc/parser/parse.rs:515]: This is amazing, it's working!!

File: crates/oxc_semantic/src/jsdoc/parser/parse.rs (hunk 1 above)
Comment (leaysgur) [reply crates/oxc_semantic/src/jsdoc/parser/parse.rs:41]:
```diff
-        while let Some((_, c)) = self.source_text[self.current..].char_indices().next() {
+        while let Some((i, c)) = self.source_text[self.current..].char_indices().next() {
             match c {
                 '@' => {
-                    self.current += c.len_utf8();
+                    self.current = i;
                     tags.push(self.parse_tag());
                 }
                 _ => {
-                    self.current += c.len_utf8();
+                    self.current = i;
                 }
             }
         }
```

I tried this, but when I did, the test hung and stopped working. What am I missing?

**Comment by Boshen** [comment]
You can improve the whole lexer by using `Chars` and `Peekable`: 

```rust
#[derive(Debug)]
pub struct JSDocParser<'a> {
    chars: Peekable<Chars<'a>>,
}

impl<'a> JSDocParser<'a> {
    pub fn new(source_text: &'a str) -> Self {
        let chars = source_text.trim().chars().peekable();
        Self { chars }
    }
```

Or maybe just `Chars` + an offset

```rust
#[derive(Debug)]
pub struct JSDocParser<'a> {
    chars: Chars<'a>,
    current: usize,
}
```

**Comment by leaysgur** [comment]
Thank you for your time + support @Boshen !!

> by using Chars and Peekable:

I will try it again one day.

**Pull request merged**